### PR TITLE
adopt TypeScript 5.5

### PR DESCRIPTION
**Pull request by turadg**
refs: #9152 

## Description

Typescript 5.5: https://devblogs.microsoft.com/typescript/announcing-typescript-5-5-rc/

(an earlier push also had https://github.com/Agoric/agoric-sdk/pull/9479 but I moved the commits out)

### Security Considerations

n/a, linting

### Scaling Considerations

n/a, linting

### Documentation Considerations

none

### Testing Considerations

none

### Upgrade Considerations

n/a, linting

### CODE REVIEW

**Comment by mhofman** [comment]
> we could change `E.when` not to return a promise in that case.

Why would we do that? Because agoric-sdk doesn't chain with `E.when` doesn't mean it's not a possible use case. Also the fact that there is a handler doesn't mean a rejection is handled. I had seen misguided suppressions like that.

**Comment by turadg** [comment]
> Why would we do that?

The paragraph you're quoting gives a reason why we would. If you're asking why _should_, I didn't claim we should. I was just spitballing.

> Because agoric-sdk doesn't chain with E.when doesn't mean it's not a possible use case. 

agoric-sdk does await E.when a handful of times. The quoted paragraph says "almost always `void`. Here's one,
```
  await E.when(outcomeOfUpdate, outcome => t.fail(`${outcome}`)).catch(e =>
    t.is(e, 'No quorum'),
  );
```

That _could_ be done with passing a promise to `.catch()` and awaiting it after E.when, but that would be very awkward.

File: .eslintrc.cjs [hunk 1]
```diff
@@ -50,7 +50,14 @@ module.exports = {
   reportUnusedDisableDirectives: true,
 
   rules: {
-    '@typescript-eslint/prefer-ts-expect-error': 'error',
+    '@typescript-eslint/ban-ts-comment': [
+      'error',
+      {
+        // TODO tigthen to 'allow-with-description'
+        'ts-expect-error': false,
+        'ts-nocheck': false,
```
Comment (mhofman) [inline .eslintrc.cjs]: Wondering if we should try to get rid of explicit `ts-nocheck` as much as possible. Maybe with some override config to disable checks wholesale where needed (e.g. cosmic-proto/codegen)

**Review by mhofman — APPROVED** [review]
LGTM

File: .eslintrc.cjs (hunk 1 above)
Comment (turadg) [reply .eslintrc.cjs]: Eventually, yeah. We have 114 files with it outside codegen. Thankfully 93 of those are tests. 14 of the non-test ones are in SwingSet. Thankfully 13 are in misc-tools.

Just 2 non-codegen `src` files, so we could it up for `src` pretty easily. Another time tho.

File: yarn.lock [hunk 2]
```diff
@@ -11584,7 +11584,12 @@ typescript-eslint@^7.3.1, typescript-eslint@^7.7.1:
   resolved "https://registry.yarnpkg.com/typescript/-/typescript-4.9.5.tgz#095979f9bcc0d09da324d58d03ce8f8374cbe65a"
   integrity sha512-1FXk9E2Hm+QzZQ7z+McJiHL4NW1F2EzMu9Nq9i3zAaGqibafqYwCVU6WyWAuyQRRzOlxou8xZSyXLEN8oKj24g==
 
-typescript@^5.5.0-beta, typescript@~5.5.0-dev.20240327:
+typescript@^5.5.1-rc:
+  version "5.5.1-rc"
+  resolved "https://registry.yarnpkg.com/typescript/-/typescript-5.5.1-rc.tgz#4be420f1ecb5e66c52a7cb11444f06f02294c977"
+  integrity sha512-7cbqBhh2QH0RaI7AD0ElJ2Ww/iRdW1w2wH/S2dv6EbdNQQlv39fx+V5VOepxLgfUvRkU5D5pxzgPuvHSAQOdpQ==
+
+typescript@~5.5.0-dev.20240327:
```
Comment (michaelfig) [inline yarn.lock:11592]: This dependency on the `dev.20240327` tag comes from `@endo/eslint-plugin`.

Would that cause problems?  If so, maybe a `packageJson.resolutions` entry is in order?  Could you consider creating an issue (or PR) on https://github.com/endojs/endo to fix those dependencies?  Waiting until we update to the official 5.5 release is also acceptable.

**Review by michaelfig — APPROVED** [review]
LGTM for the agoric-sdk.  My approval is conditional on you answering my comment in a way that satisfies you. :)

File: yarn.lock (hunk 2 above)
Comment (turadg) [reply yarn.lock:11592]: I don't think it would create problems. 

It'll line up after next sync https://github.com/endojs/endo/pull/2326

**Pull request merged**